src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -1,167 +1,373 @@
+          nextState(topicPart);
+        case WAL_CREATED:
+          applyWAL(topicPart);
+          nextState(topicPart);
+        case WAL_APPLIED:
+          truncateWAL(topicPart);
+          nextState(topicPart);
+        case WAL_TRUNCATED:
+          resetOffsets(topicPart);
+          nextState(topicPart);
+        case OFFSET_RESET:
+          resume(topicPart);
+          nextState(topicPart);
+          break;
+        default:
+          log.error("{} is not a valid state to perform recovery.", states.get(topicPart));
       }
-    } catch (IOException e) {
-      throw new CopycatRuntimeException(e);
+    } catch (CopycatException e) {
+      log.error("Recovery failed.");
+      recordFailureTime(topicPart);
+      setRetryBackoff(backOffMs);
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
-      rotate(topicPart);
+  private void execute(TopicPartition topicPart) throws IOException {
+    Queue<SinkRecord> records = buffer.get(topicPart);
+    while(!records.isEmpty()) {
+      try {
+        switch (states.get(topicPart)) {
+          case RECOVERY_STARTED:
+            pause(topicPart);
+            nextState(topicPart);
+          case RECOVERY_PARTITION_PAUSED:
+            createWAL(topicPart);
+            nextState(topicPart);
+          case WAL_CREATED:
+            applyWAL(topicPart);
+            nextState(topicPart);
+          case WAL_APPLIED:
+            truncateWAL(topicPart);
+            nextState(topicPart);
+          case WAL_TRUNCATED:
+            resetOffsets(topicPart);
+            nextState(topicPart);
+          case OFFSET_RESET:
+            resume(topicPart);
+            nextState(topicPart);
+          case WRITE_STARTED:
+            pause(topicPart);
+            nextState(topicPart);
+          case WRITE_PARTITION_PAUSED:
+            writeRecord(topicPart, records.peek());
+            records.poll();
+            if (shouldRotate(topicPart)) {
+              nextState(topicPart);
+            } else {
+              break;
+            }
+          case SHOULD_ROTATE:
+            closeTempfile(topicPart);
+            nextState(topicPart);
+          case TEMP_FILE_CLOSED:
+            appendToWAL(topicPart);
+            nextState(topicPart);
+          case WAL_APPENDED:
+            commitFile(topicPart);
+            nextState(topicPart);
+          case FILE_COMMITTED:
+            setState(topicPart, State.WRITE_PARTITION_PAUSED);
+        }
+      } catch (IllegalWorkerStateException e) {
+        // Should we retry in this case?
+        throw e;
+      } catch (IOException | CopycatException e) {
+        log.error("Exception on {}", topicPart);
+        recordFailureTime(topicPart);
+        setRetryBackoff(backOffMs);
+        break;
+      }
     }
-    writers.clear();
-    if (fs != null) {
-      fs.close();
+    if (records.isEmpty()) {
+      resume(topicPart);
```

Comment:
I think this still suffers from the same problem I described above. If we `resume()` here after exiting the loop having committed the file under the `FILE_COMMITTED` case and having transitioned to `WRITE_PARTITION_PAUSED`, we will not have paused the next time we enter this loop, right? Which seems like it violates the assumptions of all the states from `WRITE_PARTITION_PAUSED` on.

I think one way to address this would be to also reset the state to `WRITE_STARTED` when you resume (just as you do when you process the `OFFSET_RESET` state).